File: packages/contracts-ts/src/events.ts
/**
 * Agent → Shell event payloads (IPC_PROTOCOL.md §8) plus `Notification`; mirror of the "Event payloads" region of
 * `ClubShell.Contracts.Ipc.IpcMessage.cs` and the notification types of `Users/UserProfile.cs`.
 * The Shell re-emits each event as Tauri `agent://<name>` with the same payload.
 */
import type { IpcError, JsonObject, Policy, UpdateComponent, UpdateManifest, UpdatePhase } from './commands.js';
import type { GameState } from './games.js';
import type { ConnectivityState } from './pc.js';
import type { AuthExpiredReason } from './user.js';

// ---- BEGIN MANUAL ----
import { IpcNames, type LockCommand } from './commands.js';
import type { PcMetrics } from './pc.js';
import type { Session, SessionEndedEvent, SessionWarning } from './session.js';
import type { Order } from './shop.js';
import type { ChatMessage } from './user.js';
import type { Balance } from './wallet.js';
// ---- END MANUAL ----

// ---------------------------------------------------------------------------------------------------------------------
// Notifications
// ---------------------------------------------------------------------------------------------------------------------

/** Notification severity. */
export const NotificationLevel = {
  Info: 'info',
  Warning: 'warning',
  Error: 'error',
  Success: 'success',
} as const;
/** Notification severity. */
export type NotificationLevel = (typeof NotificationLevel)[keyof typeof NotificationLevel];

/** Optional call-to-action of a {@link Notification}. */
export interface NotificationAction {
  /** Button label. */
  label: string;
  /** Frontend route (e.g. `/shop`) or an `AgentCommand` IPC name (e.g. `wallet.topupIntent`). */
  command: string;
  /** Arguments for `command`. */
  args?: JsonObject | null;
}

/** Toast/notification (IPC_PROTOCOL.md §6.21); payload of `notification.push`. */
export interface Notification {
  /** Notification id. */
  id: string;
  /** Title. */
  title: string;
  /** Body text. */
  body: string;
  /** Severity. */
  level: NotificationLevel;
  /** Auto-dismiss after this many seconds; null = sticky. */
  ttlSec?: number | null;
  /** Optional call-to-action. */
  action?: NotificationAction | null;
}

// ---------------------------------------------------------------------------------------------------------------------
// Event payloads
// ---------------------------------------------------------------------------------------------------------------------

/** Remote-control state (`admin.remoteControl`). */
export const RemoteControlState = {
  Started: 'started',
  Stopped: 'stopped',
} as const;
/** Remote-control state (`admin.remoteControl`). */
export type RemoteControlState = (typeof RemoteControlState)[keyof typeof RemoteControlState];

/** UI command carried by `shell.command`. */
export const ShellCommandKind = {
  Lock: 'lock',
  Unlock: 'unlock',
  Reboot: 'reboot',
  ShowAds: 'showAds',
  ShowMessage: 'showMessage',
} as const;
/** UI command carried by `shell.command`. */
export type ShellCommandKind = (typeof ShellCommandKind)[keyof typeof ShellCommandKind];

/** Media type of an ad item. */
export const AdMediaType = {
  Image: 'image',
  Video: 'video',
} as const;
/** Media type of an ad item. */
export type AdMediaType = (typeof AdMediaType)[keyof typeof AdMediaType];

/** Payload of `admin.message`; ack via `sys.ackAdminMessage`. */
export interface AdminMessage {
  /** Message id. */
  id: string;
  /** Sender name. */
  from: string;
  /** Text. */
  text: string;
  /** Severity. */
  level: NotificationLevel;
  /** User must acknowledge (modal). */
  requiresAck: boolean;
  /** Receipt time. */
  at: string;
}

/** Payload of `admin.remoteControl`. */
export interface RemoteControlEvent {
  /** Started/stopped. */
  state: RemoteControlState;
  /** Transition time. */
  at: string;
  /** Show the on-screen indicator. */
  showIndicator: boolean;
  /** Admin name, when known. */
  adminName?: string | null;
}

/** Payload of `game.stateChanged`. */
export interface GameStateChanged {
  /** Game. */
  gameId: string;
  /** Game title. */
  title: string;
  /** New state. */
  state: GameState;
  /** Transition time. */
  at: string;
  /** Process id, when known. */
  pid?: number | null;
  /** Exit code (`exited`). */
  exitCode?: number | null;
  /** Failure (`failed`). */
  error?: IpcError | null;
}

/** Payload of `policy.changed`. */
export interface PolicyChanged {
  /** Policy version. */
  version: number;
  /** Policy timestamp. */
  updatedAt: string;
  /** Top-level sections that changed. */
  changed: string[];
  /** Policy now in effect. */
  policy: Policy;
}

/** Payload of `update.available`. */
export interface UpdateAvailable {
  /** Newer package. */
  manifest: UpdateManifest;
  /** Installed version of that component. */
  current: string;
}

/** Payload of `update.progress` (≤ 2/s). */
export interface UpdateProgress {
  /** Component. */
  component: UpdateComponent;
  /** Package version. */
  version: string;
  /** Phase. */
  phase: UpdatePhase;
  /** 0–100. */
  percent: number;
  /** Bytes processed. */
  bytesDone: number;
  /** Total bytes. */
  bytesTotal: number;
  /** Failure (`failed`). */
  error?: IpcError | null;
}

/** Payload of `update.ready` (package staged and verified). */
export interface UpdateReady {
  /** Component. */
  component: UpdateComponent;
  /** Package version. */
  version: string;
  /** Whether applying restarts the component. */
  restartRequired: boolean;
  /** Will be applied even during a session. */
  mandatory: boolean;
  /** Scheduled apply time, when known. */
  applyAt?: string | null;
}

/** Payload of `sys.connectivity` (transitions + every 60 s). */
export interface ConnectivityEvent {
  /** Connectivity. */
  state: ConnectivityState;
  /** When the state was entered. */
  since: string;
  /** Outbox size. */
  queuedEvents: number;
  /** Last heartbeat latency, when online. */
  serverLatencyMs?: number | null;
}

/** Args of `shell.command{reboot}` (informational; the Agent performs the reboot). */
export interface ShellRebootArgs {
  /** Seconds until reboot. */
  delaySec: number;
  /** Message to display. */
  message?: string | null;
}

/** One ad in a {@link ShowAdsArgs} playlist. */
export interface AdItem {
  /** Media URL. */
  url: string;
  /** Media type. */
  type: AdMediaType;
  /** Display duration. */
  durationSec: number;
}

/** Args of `shell.command{showAds}` and payload of the `showAds` server command. */
export interface ShowAdsArgs {
  /** Playlist. */
  items: AdItem[];
  /** User may skip. */
  skippable: boolean;
}

/** Args of `shell.command{showMessage}`. */
export interface ShowMessageArgs {
  /** Title. */
  title: string;
  /** Body. */
  body: string;
  /** Severity. */
  level: NotificationLevel;
  /** Auto-dismiss after this many seconds. */
  ttlSec?: number | null;
}

// ---- BEGIN MANUAL ----
/** Typed `args` per {@link ShellCommandKind}. */
export interface ShellCommandArgsMap {
  lock: LockCommand | null;
  unlock: null;
  reboot: ShellRebootArgs;
  showAds: ShowAdsArgs;
  showMessage: ShowMessageArgs;
}

/** Payload of `shell.command`; discriminated on `command`, `args` key always present. */
export type ShellCommand = {
  [K in ShellCommandKind]: {
    /** Command kind. */
    command: K;
    /** Typed args or null. */
    args: ShellCommandArgsMap[K];
    /** Originating server command id (or a fresh id for Agent-originated commands). */
    commandId: string;
  };
}[ShellCommandKind];
// ---- END MANUAL ----

/** Payload of `auth.expired`; the Shell returns to the login screen. */
export interface AuthExpired {
  /** Why the user context was invalidated. */
  reason: AuthExpiredReason;
}

// ---------------------------------------------------------------------------------------------------------------------
// Event name → payload
// ---------------------------------------------------------------------------------------------------------------------

// ---- BEGIN MANUAL ----
/** Payload type per Agent → Shell event name (IPC_PROTOCOL.md §8). */
export interface AgentEventMap {
  [IpcNames.Events.SessionUpdated]: Session;
  [IpcNames.Events.SessionWarning]: SessionWarning;
  [IpcNames.Events.SessionEnded]: SessionEndedEvent;
  [IpcNames.Events.WalletUpdated]: Balance;
  [IpcNames.Events.ChatMessage]: ChatMessage;
  [IpcNames.Events.NotificationPush]: Notification;
  [IpcNames.Events.AdminMessage]: AdminMessage;
  [IpcNames.Events.AdminRemoteControl]: RemoteControlEvent;
  [IpcNames.Events.GameStateChanged]: GameStateChanged;
  [IpcNames.Events.PolicyChanged]: PolicyChanged;
  [IpcNames.Events.UpdateAvailable]: UpdateAvailable;
  [IpcNames.Events.UpdateProgress]: UpdateProgress;
  [IpcNames.Events.UpdateReady]: UpdateReady;
  [IpcNames.Events.SysMetrics]: PcMetrics;
  [IpcNames.Events.SysConnectivity]: ConnectivityEvent;
  [IpcNames.Events.ShellCommand]: ShellCommand;
  [IpcNames.Events.AuthExpired]: AuthExpired;
  [IpcNames.Events.ShopOrderUpdated]: Order;
}

/** Agent → Shell event name (`'session.updated' | 'session.warning' | ...`). */
export type AgentEventName = keyof AgentEventMap;

/** Discriminated union of every Agent → Shell event: `{ name, payload }`. */
export type ShellEvent = {
  [K in AgentEventName]: {
    /** Event name. */
    name: K;
    /** Typed payload. */
    payload: AgentEventMap[K];
  };
}[AgentEventName];

/** All Agent → Shell event names in declaration order. */
export const AGENT_EVENT_NAMES: readonly AgentEventName[] = Object.values(IpcNames.Events);

const AGENT_EVENT_NAME_SET: ReadonlySet<string> = new Set<string>(AGENT_EVENT_NAMES);

/** `true` when `name` is a known Agent → Shell event name. */
export function isAgentEventName(name: string): name is AgentEventName {
  return AGENT_EVENT_NAME_SET.has(name);
}
// ---- END MANUAL ----

File: crates/protocol/src/events.rs
//! Agent → Shell event payloads (IPC_PROTOCOL.md §8; C# `IpcMessage.cs` "Event payloads" region).
//! Payloads defined next to their domain are re-exported here so every event payload is reachable
//! from this module: [`SessionWarning`], [`SessionEndedEvent`], [`SessionEvent`], [`Notification`].
//! Names live in [`names::events`](crate::commands::names::events).

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;

use crate::commands::{UpdateComponent, UpdateManifest, UpdatePhase};
use crate::error::IpcError;
use crate::games::GameState;
use crate::pc::{ConnectivityState, Policy};
use crate::user::{AuthExpiredReason, NotificationLevel};

// ---- BEGIN MANUAL ----
use serde::de::DeserializeOwned;

use crate::commands::{object_as, MessageCommand};
use crate::error::ProtocolError;

pub use crate::commands::names::events as event_names;
pub use crate::commands::LockCommand;
pub use crate::session::{SessionEndedEvent, SessionEvent, SessionEventType, SessionWarning};
pub use crate::user::{Notification, NotificationAction};
// ---- END MANUAL ----

wire_enum! {
    /// Remote-control state (`admin.remoteControl`).
    RemoteControlState {
        Started = "started",
        Stopped = "stopped",
    }
}

wire_enum! {
    /// UI command carried by `shell.command`.
    ShellCommandKind {
        /// Show the lock screen; args [`LockCommand`].
        Lock = "lock",
        /// Hide the lock screen; no args.
        Unlock = "unlock",
        /// Reboot notice (Agent performs the reboot); args [`ShellRebootArgs`].
        Reboot = "reboot",
        /// Play ads; args [`ShowAdsArgs`].
        ShowAds = "showAds",
        /// Show a message; args [`ShowMessageArgs`].
        ShowMessage = "showMessage",
    }
}

wire_enum! {
    /// Media type of an ad item.
    AdMediaType {
        Image = "image",
        Video = "video",
    }
}

/// Payload of `admin.message`.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AdminMessage {
    /// Ack via `sys.ackAdminMessage`.
    pub id: Uuid,
    /// Sender name.
    pub from: String,
    pub text: String,
    pub level: NotificationLevel,
    /// User must acknowledge (modal).
    pub requires_ack: bool,
    /// Receipt time.
    #[serde(with = "crate::wire::ts")]
    pub at: DateTime<Utc>,
}

// ---- BEGIN MANUAL ----
impl AdminMessage {
    /// Builds from the server command payload.
    pub fn from_command(command: &MessageCommand, at: DateTime<Utc>) -> Self {
        Self {
            id: command.id,
            from: command.from.clone(),
            text: command.text.clone(),
            level: command.level,
            requires_ack: command.requires_ack,
            at,
        }
    }
}
// ---- END MANUAL ----

/// Payload of `admin.remoteControl`.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RemoteControlEvent {
    pub state: RemoteControlState,
    #[serde(with = "crate::wire::ts")]
    pub at: DateTime<Utc>,
    /// Show the on-screen indicator.
    pub show_indicator: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub admin_name: Option<String>,
}

/// Payload of `game.stateChanged`.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct GameStateChanged {
    pub game_id: Uuid,
    pub title: String,
    pub state: GameState,
    #[serde(with = "crate::wire::ts")]
    pub at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pid: Option<i32>,
    /// Exit code (`exited`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub exit_code: Option<i32>,
    /// Failure (`failed`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<IpcError>,
}

/// Payload of `policy.changed`.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PolicyChanged {
    pub version: i32,
    #[serde(with = "crate::wire::ts")]
    pub updated_at: DateTime<Utc>,
    /// Top-level sections that changed.
    pub changed: Vec<String>,
    pub policy: Policy,
}

/// Payload of `update.available`.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct UpdateAvailable {
    pub manifest: UpdateManifest,
    /// Installed version of that component.
    pub current: String,
}

/// Payload of `update.progress` (≤ 2/s).
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct UpdateProgress {
    pub component: UpdateComponent,
    pub version: String,
    pub phase: UpdatePhase,
    /// 0–100.
    pub percent: i32,
    pub bytes_done: i64,
    pub bytes_total: i64,
    /// Failure (`failed`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<IpcError>,
}

/// Payload of `update.ready` (package staged and verified).
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct UpdateReady {
    pub component: UpdateComponent,
    pub version: String,
    /// Whether applying restarts the component.
    pub restart_required: bool,
    /// Will be applied even during a session.
    pub mandatory: bool,
    #[serde(default, with = "crate::wire::ts_opt", skip_serializing_if = "Option::is_none")]
    pub apply_at: Option<DateTime<Utc>>,
}

/// Payload of `sys.connectivity` (transitions + every 60 s).
#[derive(Serialize, Deserialize, Clone, Copy, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ConnectivityEvent {
    pub state: ConnectivityState,
    /// When the state was entered.
    #[serde(with = "crate::wire::ts")]
    pub since: DateTime<Utc>,
    /// Outbox size.
    pub queued_events: i32,
    /// Last heartbeat latency, when online.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub server_latency_ms: Option<i32>,
}

/// Args of [`ShellCommandKind::Reboot`] (informational; the Agent performs the reboot).
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ShellRebootArgs {
    pub delay_sec: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

/// One ad in a [`ShowAdsArgs`] playlist.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AdItem {
    pub url: String,
    pub r#type: AdMediaType,
    pub duration_sec: i32,
}

/// Args of [`ShellCommandKind::ShowAds`] and payload of the `showAds` server command.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ShowAdsArgs {
    pub items: Vec<AdItem>,
    pub skippable: bool,
}

/// Args of [`ShellCommandKind::ShowMessage`].
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ShowMessageArgs {
    pub title: String,
    pub body: String,
    pub level: NotificationLevel,
    /// Auto-dismiss after this many seconds.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ttl_sec: Option<i32>,
}

/// Payload of `shell.command`. `args` is always present on the wire (`null` when the command has none).
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ShellCommand {
    pub command: ShellCommandKind,
    /// Typed args (see [`ShellCommandKind`]) or `None`.
    pub args: Option<Value>,
    /// Originating server command id (or a fresh id for Agent-originated commands).
    pub command_id: Uuid,
}

// ---- BEGIN MANUAL ----
impl ShellCommand {
    /// Command with typed args.
    pub fn of<T: Serialize>(command: ShellCommandKind, args: Option<&T>, command_id: Uuid) -> Result<Self, ProtocolError> {
        let args = match args {
            Some(a) => Some(serde_json::to_value(a)?),
            None => None,
        };
        Ok(Self { command, args, command_id })
    }

    /// Deserializes `args` as `T`; `Ok(None)` when absent or not an object.
    pub fn args_as<T: DeserializeOwned>(&self) -> Result<Option<T>, ProtocolError> {
        object_as(self.args.as_ref())
    }
}
// ---- END MANUAL ----

/// Payload of `auth.expired`; the Shell returns to the login screen.
#[derive(Serialize, Deserialize, Clone, Copy, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AuthExpired {
    pub reason: AuthExpiredReason,
}

// ---- BEGIN MANUAL ----
#[cfg(test)]
mod tests {
    use super::*;
    use crate::commands::{UpdateChannel, UpdateManifest};
    use crate::error::{assert_wire, ErrorCode};
    use chrono::TimeZone;

    fn t(h: u32, m: u32, s: u32) -> DateTime<Utc> {
        Utc.with_ymd_and_hms(2026, 9, 21, h, m, s).unwrap()
    }

    #[test]
    fn enum_wire_values() {
        assert_wire(RemoteControlState::ALL, &["started", "stopped"]);
        assert_wire(ShellCommandKind::ALL, &["lock", "unlock", "reboot", "showAds", "showMessage"]);
        assert_wire(AdMediaType::ALL, &["image", "video"]);
    }

    #[test]
    fn shell_command_keeps_null_args() {
        let cmd = ShellCommand::of::<LockCommand>(ShellCommandKind::Unlock, None, Uuid::nil()).unwrap();
        assert_eq!(
            serde_json::to_string(&cmd).unwrap(),
            r#"{"command":"unlock","args":null,"commandId":"00000000-0000-0000-0000-000000000000"}"#
        );
        assert_eq!(cmd.args_as::<LockCommand>().unwrap(), None);
        let lock = LockCommand { reason: Some("admin".into()), message: Some("Please come to the desk".into()) };
        let cmd = ShellCommand::of(ShellCommandKind::Lock, Some(&lock), Uuid::nil()).unwrap();
        let json = serde_json::to_string(&cmd).unwrap();
        assert_eq!(
            json,
            r#"{"command":"lock","args":{"reason":"admin","message":"Please come to the desk"},"commandId":"00000000-0000-0000-0000-000000000000"}"#
        );
        let back: ShellCommand = serde_json::from_str(&json).unwrap();
        assert_eq!(back.args_as::<LockCommand>().unwrap(), Some(lock));
        let ads = ShowAdsArgs { items: vec![AdItem { url: "https://a.mp4".into(), r#type: AdMediaType::Video, duration_sec: 15 }], skippable: true };
        assert_eq!(serde_json::to_string(&ads).unwrap(), r#"{"items":[{"url":"https://a.mp4","type":"video","durationSec":15}],"skippable":true}"#);
        let msg = ShowMessageArgs { title: "t".into(), body: "b".into(), level: NotificationLevel::Info, ttl_sec: None };
        assert_eq!(serde_json::to_string(&msg).unwrap(), r#"{"title":"t","body":"b","level":"info"}"#);
        assert_eq!(serde_json::to_string(&ShellRebootArgs { delay_sec: 30, message: None }).unwrap(), r#"{"delaySec":30}"#);
    }

    #[test]
    fn admin_and_remote_control_events() {
        let cmd = MessageCommand { id: Uuid::nil(), from: "Admin".into(), text: "hi".into(), level: NotificationLevel::Warning, requires_ack: true };
        let m = AdminMessage::from_command(&cmd, t(10, 0, 0));
        let json = serde_json::to_string(&m).unwrap();
        assert_eq!(
            json,
            r#"{"id":"00000000-0000-0000-0000-000000000000","from":"Admin","text":"hi","level":"warning","requiresAck":true,"at":"2026-09-21T10:00:00.000Z"}"#
        );
        assert_eq!(serde_json::from_str::<AdminMessage>(&json).unwrap(), m);
        let rc = RemoteControlEvent { state: RemoteControlState::Started, at: t(10, 0, 0), show_indicator: true, admin_name: Some("Bob".into()) };
        assert_eq!(
            serde_json::to_string(&rc).unwrap(),
            r#"{"state":"started","at":"2026-09-21T10:00:00.000Z","showIndicator":true,"adminName":"Bob"}"#
        );
        assert_eq!(serde_json::to_string(&AuthExpired { reason: AuthExpiredReason::Revoked }).unwrap(), r#"{"reason":"revoked"}"#);
        let conn = ConnectivityEvent { state: ConnectivityState::Offline, since: t(9, 0, 0), queued_events: 4, server_latency_ms: None };
        assert_eq!(serde_json::to_string(&conn).unwrap(), r#"{"state":"offline","since":"2026-09-21T09:00:00.000Z","queuedEvents":4}"#);
    }

    #[test]
    fn game_policy_update_events() {
        let g = GameStateChanged { game_id: Uuid::nil(), title: "CS2".into(), state: GameState::Failed, at: t(10, 21, 0), pid: None, exit_code: None, error: Some(IpcError::of(ErrorCode::GameNotInstalled)) };
        let json = serde_json::to_string(&g).unwrap();
        assert_eq!(
            json,
            r#"{"gameId":"00000000-0000-0000-0000-000000000000","title":"CS2","state":"failed","at":"2026-09-21T10:21:00.000Z","error":{"code":"gameNotInstalled","message":"Game is not installed","details":null}}"#
        );
        assert_eq!(serde_json::from_str::<GameStateChanged>(&json).unwrap(), g);

        let p = PolicyChanged { version: 12, updated_at: t(8, 0, 0), changed: vec!["kiosk".into()], policy: crate::pc::sample_policy() };
        let json = serde_json::to_string(&p).unwrap();
        assert!(json.starts_with(r#"{"version":12,"updatedAt":"2026-09-21T08:00:00.000Z","changed":["kiosk"],"policy":{"version":12"#));
        assert_eq!(serde_json::from_str::<PolicyChanged>(&json).unwrap(), p);

        let manifest = UpdateManifest {
            channel: UpdateChannel::Beta,
            component: UpdateComponent::Agent,
            version: "1.5.0".into(),
            url: "https://u".into(),
            sha256: "ab".into(),
            size: 1,
            signature: "s".into(),
            release_notes: String::new(),
            mandatory: true,
            published_at: t(3, 0, 0),
            min_agent_version: None,
        };
        let avail = UpdateAvailable { manifest, current: "1.4.2".into() };
        let json = serde_json::to_string(&avail).unwrap();
        assert!(json.starts_with(r#"{"manifest":{"channel":"beta","component":"agent","version":"1.5.0""#));
        assert!(json.ends_with(r#""publishedAt":"2026-09-21T03:00:00.000Z"},"current":"1.4.2"}"#));
        let progress = UpdateProgress { component: UpdateComponent::Agent, version: "1.5.0".into(), phase: UpdatePhase::Downloading, percent: 42, bytes_done: 4200, bytes_total: 10000, error: None };
        assert_eq!(
            serde_json::to_string(&progress).unwrap(),
            r#"{"component":"agent","version":"1.5.0","phase":"downloading","percent":42,"bytesDone":4200,"bytesTotal":10000}"#
        );
        let ready = UpdateReady { component: UpdateComponent::Shell, version: "1.5.0".into(), restart_required: true, mandatory: false, apply_at: Some(t(4, 0, 0)) };
        assert_eq!(
            serde_json::to_string(&ready).unwrap(),
            r#"{"component":"shell","version":"1.5.0","restartRequired":true,"mandatory":false,"applyAt":"2026-09-21T04:00:00.000Z"}"#
        );
        assert_eq!(event_names::ALL.len(), 18);
    }
}
// ---- END MANUAL ----
